src/Listeners/QueueJobSubscriber.php

```diff
@@ -0,0 +1,161 @@
+<?php
+
+namespace Vinelab\Tracing\Listeners;
+
+use Illuminate\Container\Container;
+use Illuminate\Contracts\Support\Arrayable;
+use Illuminate\Contracts\Support\Jsonable;
+use Illuminate\Queue\Events\JobFailed;
+use Illuminate\Queue\Events\JobProcessed;
+use Illuminate\Queue\Events\JobProcessing;
+use Illuminate\Queue\Jobs\Job;
+use Illuminate\Support\Arr;
+use Illuminate\Support\Collection;
+use JsonSerializable;
+use Psr\Container\ContainerInterface;
+use ReflectionClass;
+use ReflectionParameter;
+use Vinelab\Tracing\Contracts\ShouldBeTraced;
+use Vinelab\Tracing\Contracts\SpanContext;
+use Vinelab\Tracing\Facades\Trace;
+
+class QueueJobSubscriber
+{
+    /**
+     * @var ContainerInterface
+     */
+    protected $app;
+
+    /**
+     * QueueJobSubscriber constructor.
+     * @param  Container  $app
+     */
+    public function __construct(Container $app)
+    {
+        $this->app = $app;
+    }
+
+    /**
+     * @param  JobProcessing  $event
+     */
+    public function onJobProcessing(JobProcessing $event)
+    {
+        $this->startQueueJobSpan($event);
+    }
+
+    /**
+     * @param  JobProcessed  $event
+     */
+    public function onJobProcessed(JobProcessed $event)
+    {
+        $this->closeQueueJobSpan($event);
+    }
+
+    /**
+     * @param  JobFailed  $event
+     */
+    public function onJobFailed(JobFailed $event)
+    {
+        $this->closeQueueJobSpan($event);
+    }
+
+    /**
+     * @param  JobProcessing  $event
+     */
+    protected function startQueueJobSpan(JobProcessing $event)
+    {
+        /** @var Job $job */
+        $job = $event->job;
+        /** @var mixed $jobInstance */
+        $jobInstance = unserialize(Arr::get($job->payload(), 'data.command'));
+
+        if (in_array(ShouldBeTraced::class, class_implements($jobInstance))) {
+            $jobInput = $this->retrieveQueueJobInput($jobInstance);
+
+            $span = Trace::startSpan(class_basename($job->resolveName()), $jobInput->first(function ($attr) {
```

Comment:
Is the class name a relevan name for the span? I would keep it as standard as possible.

Reply:
In Laravel queue system, yes. The next best thing is the name of the queue and that can be pretty generic depending on how application is structured. It's not uncommon to have jobs of different type being processed on the same queue (i.e. `default`) in Laravel. Which is why the queue is recorded as a tag, but class name is used as the span name.